### ISSUE TITLE
Different class name in spatialreg seems to be breaking piecewiseSEM

### ISSUE THREAD

**Issue by frandadamo** [jslefche/piecewiseSEM#247]
Hi everyone,

I am trying to account for spatial autocorrelation within a SEM using Jon's piecewiseSEM.

[Not too long ago](https://github.com/r-spatial/spatialreg/issues/2), the spdep package was split into two packages, i.e., spded and spatialreg. Before this splitting, psem accounting for spatial autocorrelation run ok (as far as I am aware). Now, this seems not to be happening anymore. It seems that the issue is likely related to different class names in the new version of the spded/spatialreg packages (see other GitHub issues, e.g., [here](https://github.com/leifeld/texreg/issues/180), [here](https://github.com/tidymodels/broom/issues/1003), #203, etc.).

Below is a reprex using a small chunk of my data:

```
library(piecewiseSEM)
library(spatialreg)
library(dplyr)
library(spdep)

# #reproduce just 10 rows from my data
mydata <- data.frame(lat=c(30.12,29.88,30.12,29.12,29.88,30.12,25.88,26.12,26.38,26.62),
                     long=c(-25.12,-25.38,-25.38,-25.62,-25.62,-25.62,-25.88,-25.88,-25.88,-25.88),
                     resp1=c(0.6130,0.5990,0.6410,0.5318,0.5551,0.6280,0.4611,0.5046,0.5400,0.5491),
                     resp2=c(23.04,22.69,20.93,24.15,18.74,21.28,17.01,15.20,16.08,16.68),
                     ind1=c(0.6159,0.5782,0.6339,0.5313,0.5774,0.6102,0.4174,0.4816,0.5114,0.5269),
                     ind2=c(1674,1351,1618,1518,1311,1483,1167,1199,1255,1380),
                     ind3=c(917.4,736.7,896.5,735.2,714.7,800.9,589.8,613.1,662.1,743.5))

# #create spatial weighting matrix
dfxy  <- mydata %>% dplyr::select(lat,long)
nb_my <- tri2nb(dfxy)
plot(nb_my, dfxy, col="green", pch=20, cex=1)
listwgab <- nb2listw(nb_my)

# #try single lagsarlm
m1 <- spatialreg::lagsarlm(resp2 ~ ind1 + ind2, mydata, listwgab, tol.solve=1e-11)
summary(m1)

# #now lagsarlm within a psem
sem1 <- psem(
  spatialreg::lagsarlm(resp2 ~ ind1 + ind2, mydata, listwgab, tol.solve=1e-11),
  spatialreg::lagsarlm(resp1 ~ resp2 + ind3, mydata, listwgab, tol.solve=1e-11),
  data = mydata)
summary(sem1)
```

m1 is a single model that works fine.

sem1 does not work. The error message is:
`Error: Unsupported model class in model list: Sarlm. See 'help(piecewiseSEM)' for more details.`
However, the piecewiseSEM documentation says that sarlm is supported, so it may be just a problem related to the capitalization of the 's' ('sarlm' vs. 'Sarlm')?

Any idea? I hope this question is not too easy or silly.

Many thanks.

**Comment by pavel-fibich** [comment]
I am having similar problem, digging in piecewiseSEM code I found that it assumes class name sarlm, but new spatialreg produces Sarlm, that makes problems in several function eg. internal GetData() or getCoeficients(), but there are just up to 5 occurrences in the code.
I tried to downgrade spatialreg 1.1-5 and also spdep, it is no ideal (there some error messages during the loading of package), but few first psem's with errorsarlm produce sound results.
Please upgrade piecewiseSEM package to consider Sarlm too.

**Comment by frandadamo** [comment]
Hi Pavel,
That sounds cool. How did you downgrade spatialreg 1.1-5 and spdep?
By the way yes, Sarlm class hasn't been implemented yet (Jon said this is something he should be able to do sometime in the next few months).

**Comment by pavel-fibich** [comment]
Hi Francesco.
One way is to downgrade spatialreg and spdep:
install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.1-5.tar.gz", repos=NULL, type="source")
install.packages("https://cran.r-project.org/src/contrib/Archive/spatialreg/spatialreg_1.1-5.tar.gz", repos=NULL, type="source")
On these versions R packages produce sarlm class that 2.1.2 piecewiseSEM accepts.

Or leave spatialreg and spdep in actual versions and use my fork of 2.1.2 piecewiseSEM that accepts Sarlm class:
devtools::install_github("pavel-fibich/piecewiseSEM")
I only added "Sarlm" in the code right besides original "sarlm" (hoping that Sarlm can be used in the same way as sarlm and that sarlm is already correctly supported). I also tried to use devel branch first, but there was some problem that I was not able to solve fast.

**Comment by pavel-fibich** [comment]
Doing several tests suggest that the first way (downgrading spatialreg and spdep) + downgrading piecewiseSEM
install.packages("https://cran.r-project.org/src/contrib/Archive/piecewiseSEM/piecewiseSEM_2.0.2.tar.gz", repos=NULL, type="source")
provides solution for me. Using my fork of piecewiseSEM led to problems with summary and coefficients of fitted SEMs. I will try to investigate it later, I still have plans to use errorsarlm in psem.

**Comment by frandadamo** [comment]
Hi Pavel,
Great, thanks a lot for sharing this.
Following your suggestions I also get the summary statistics from the psem.
Just one quick question: why do you think we need to downgrade also spatialreg? I mean, downgrading spdep means using the old version, i.e., when all the functions were in spdep and spatialreg was not developed yet. Right?

**Comment by pavel-fibich** [comment]
Hi Francesco, regarding downgrading of spatialreg you may be right, I did not study it much, I just think that they somehow share versioning. Meantime I worked on my fork of piecewiseSEM and solved two problems with Sarlm models (I included Nagelkerke psedu R2, correct the error of not working terms() on Sarlm class), that leads to desired setup where I have recent spatialreg, spdep and my fork piecewiseSEM working all together (ie. producing Fishers C and summary). So I will stay on my version of piecewiseSEM. My fork can be installed by
devtools::install_github("pavel-fibich/piecewiseSEM")

**Comment by pavel-fibich** [comment]
Hi Francesco, thank you for testing.
Regarding DF in Sarlm, in the original code for sarlm (in coefs.R), DF is set to NA, not sure why, I did not change it. But I guess DF can be computed by classical formula n-k, where n is the number of observations and k is the number of parameters (do not forget to add parameters of spatial auto correlation). Calling anova on Sarlm produces some DF too.

**Comment by frandadamo** [comment]
Brilliant, thanks a lot :)!

**Comment by Tao-Liang** [comment]
Hi. May I join you? I also met s similar situation(I struggled for five months). And thanks to your comments. I know what happened now. 

But when I use pavel's forked psem (with the current versions of spatialreg&spdep ), I still get nothing{Error in as.list.default(X) :  no method for coercing this S4 class to a vector}

However, When I used downgraded version, I need to use spdep 1.1-3， and piecewiseSEM 2.0.0, then will I get a results, and most strange is when I use summary(psem), I found  Std.Estimate were NA for every model(but there are some ** for some models), I also cannot plot the psem results when I use plot(psem)(Error in xy.coords(x, y, xlabel, ylabel, log) : 
  'x' is a list, but does not have components 'x' and 'y').
Is that because I did something wrong? or? Could you please help me have a look?
Many thanks!
Tao

**Comment by Tao-Liang** [comment]
> Hi. May I join you? I also met s similar situation(I struggled for five months). And thanks to your comments. I know what happened now.
> 
> But when I use pavel's forked psem (with the current versions of spatialreg&spdep ), I still get nothing{Error in as.list.default(X) : no method for coercing this S4 class to a vector}
> 
> However, When I used downgraded version, I need to use spdep 1.1-3， and piecewiseSEM 2.0.0, then will I get a results, and most strange is when I use summary(psem), I found Std.Estimate were NA for every model(but there are some ** for some models), I also cannot plot the psem results when I use plot(psem)(Error in xy.coords(x, y, xlabel, ylabel, log) : 'x' is a list, but does not have components 'x' and 'y'). Is that because I did something wrong? or? Could you please help me have a look? Many thanks! Tao

**Comment by pavel-fibich** [comment]
Hi Tao, I guess you may have some mixed versions environment. Check which class is produced by lagsarlm function, sometimes closing/opening Rstudio helps. I am using my fork of piecewiseSEM (fork of 2.1.2 version) and spatialreg of 1.2-1 version, and they work nice together. It will be easier to guess if you provide more details about your setup.

**Comment by Tao-Liang** [comment]
Thnaks so much for your reply, dear pavel.
  Yes, I used the fork package by devtools::install_github("pavel-fibich/piecewiseSEM"). And install.packages("spatialreg"), install.packages("spdep").
  My R is version R4.1.1. And I used R studio 2021.09.0+351.
    All_237_all_variable<-read.csv("DATA/output.csv",header=T)
    coordinates(All_237_all_variable) <- ~Long+Lat##library(sp)
    All_237_nb <- spdep::tri2nb(All_237_all_variable)
    All_237_spatial_weights <- spdep::nb2listw(All_237_nb)
    All_237_model1 <- spatialreg::lagsarlm(SR ~mDR+TemVar+PreVar+MBL,   data = All_237_all_variable,listw = All_237_spatial_weights,tol.solve = 1e-11)
    All_237_model2 <- spatialreg::lagsarlm(MBL ~ TemVar+PreVar,data = All_237_all_variable,listw = All_237_spatial_weights,tol.solve = 1e-11)
    All_237_model3 <- spatialreg::lagsarlm(mDR ~ TemVar+PreVar, data = All_237_all_variable,    listw = All_237_spatial_weights,    tol.solve = 1e-11)

    All_237_psem <-piecewiseSEM::psem(  All_237_model1,  All_237_model2,  All_237_model3,  data = All_237_all_variable)

        Error in `.rowNamesDF<-`(x, value = value) : 'row.names' length incorrect
Further, I used downgrading spdep 1.2-1  &1.1-13 &1.1-12 &1.1-11 &1.1-8
   install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.2-1.tar.gz", repos=NULL, type="source")
   install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.1-13.tar.gz", repos=NULL, type="source")
   install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.1-12.tar.gz", repos=NULL, type="source")
 and when I used pse(), All got: error.
   Error in as.list.default(X) :   no method for coercing this S4 class to a vector
  I donot know what happened. Thank you in advance












At 2022-02-04 22:33:58, "Pavel Fibich" ***@***.***> wrote:

Hi Tao, I guess you may have some mixed versions environment. Check which class is produced by lagsarlm function, sometimes closing/opening Rstudio helps. I am using my fork of piecewiseSEM (fork of 2.1.2 version) and spatialreg of 1.2-1 version, and they work nice together. It will be easier to guess if you provide more details about your setup.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Tao-Liang** [comment]
> Hi Tao, I guess you may have some mixed versions environment. Check which class is produced by lagsarlm function, sometimes closing/opening Rstudio helps. I am using my fork of piecewiseSEM (fork of 2.1.2 version) and spatialreg of 1.2-1 version, and they work nice together. It will be easier to guess if you provide more details about your setup.



> Thnaks so much for your reply, dear pavel. Yes, I used the fork package by devtools::install_github("pavel-fibich/piecewiseSEM"). And install.packages("spatialreg"), install.packages("spdep"). My R is version R4.1.1. And I used R studio 2021.09.0+351. All_237_all_variable<-read.csv("DATA/output.csv",header=T) coordinates(All_237_all_variable) <- ~Long+Lat##library(sp) All_237_nb <- spdep::tri2nb(All_237_all_variable) All_237_spatial_weights <- spdep::nb2listw(All_237_nb) All_237_model1 <- spatialreg::lagsarlm(SR ~mDR+TemVar+PreVar+MBL, data = All_237_all_variable,listw = All_237_spatial_weights,tol.solve = 1e-11) All_237_model2 <- spatialreg::lagsarlm(MBL ~ TemVar+PreVar,data = All_237_all_variable,listw = All_237_spatial_weights,tol.solve = 1e-11) All_237_model3 <- spatialreg::lagsarlm(mDR ~ TemVar+PreVar, data = All_237_all_variable, listw = All_237_spatial_weights, tol.solve = 1e-11) All_237_psem <-piecewiseSEM::psem( All_237_model1, All_237_model2, All_237_model3, data = All_237_all_variable) Error in `.rowNamesDF<-`(x, value = value) : 'row.names' length incorrect Further, I used downgrading spdep 1.2-1 &1.1-13 &1.1-12 &1.1-11 &1.1-8 install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.2-1.tar.gz", repos=NULL, type="source") install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.1-13.tar.gz", repos=NULL, type="source") install.packages("https://cran.r-project.org/src/contrib/Archive/spdep/spdep_1.1-12.tar.gz", repos=NULL, type="source") and when I used pse(), All got: error. Error in as.list.default(X) : no method for coercing this S4 class to a vector I donot know what happened. Thank you in advance At 2022-02-04 22:33:58, "Pavel Fibich" ***@***.***> wrote: Hi Tao, I guess you may have some mixed versions environment. Check which class is produced by lagsarlm function, sometimes closing/opening Rstudio helps. I am using my fork of piecewiseSEM (fork of 2.1.2 version) and spatialreg of 1.2-1 version, and they work nice together. It will be easier to guess if you provide more details about your setup. — Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you commented.Message ID: ***@***.***>

OK. Finally, I get the results using Pavel  forked piecewiseSEM with 1.2-1 spatialreg.
And I Maybe know why it doesnot work before;
Because my raw csv data (I called Raw.all ) including many columns that do not used in the models.
So the only solution is prepare a new csv  (I called Raw.A) data, only have long, Lat and varibles that will be used  in the models. And read.csv again.
Creating a new csv (Raw.A) is important. Because it seems:
When I used Raw.A <- Raw.all[,c(1,2,.....)]# to chose the columns that I will use in the model.
It still doesnot work.
But if you use wirte.csv(Raw.A,"/Raw.A.csv") to save it as a csv data in the folder
And re-read it again ( Raw.A<-read.csv("/Raw.A.csv",heaer=T)),
And I get the results now.
Thanks dear Pavel~

**Comment by frandadamo** [comment]
Hi Tao,
Yes, I also have Pavel's fork with piecewiseSEM_2.1.2, spdep_1.2-1, and spatialreg_1.2-1.
A hint of Pavel's fork being implemented correctly may consist in checking if you get the Nagelkerke R2 at the bottom of the model summary.

**Comment by frandadamo** [comment]
Related to this issue, [this](https://link.springer.com/article/10.1007/s00442-022-05123-w) recent paper is quite helpful.

**Comment by frandadamo** [comment]
Hi @pavel-fibich,
Sorry I'd have a further clarification to ask you if possible, please.
While your fork does work well, I notice that it masks some functions of the piecewiseSEM package. For instance, the summary provides only the global Fisher's C statistic and does not provide the global Chi-square statistic. Similarly, the `fisherC()` function does work while the `LLchisq()` does not work. Following [this](https://github.com/jslefche/sem_book/issues/3), I (re)installed the development version of piecewiseSEM that, unfortunately, brings back the initial problem of "sarlm" vs. "Sarlm". I am aware that the Chi-square statistic can be retrieved in other ways, but I was wondering if you or @Tao-Liang had the same problem and if it would be maybe possible to address this.
Thanks a lot

**Comment by pavel-fibich** [comment]
Hi Francesco, I did not have much time to use my fork, but I must finish analyses applying it during next 2 months. May be the best approach will be to fork & improve devel version. I will try to check LLchisq() and make it working too. I will let you know.

**Comment by frandadamo** [comment]
Dear Pavel, many thanks for your message.
No problem at all. In the meanwhile, I am calculating Chi-squared manually (e.g., [here](https://jslefche.github.io/sem_book/local-estimation.html#model-fitting-using-piecewisesem) or [here](https://link.springer.com/article/10.1007/s00442-022-05123-w)).
I also noticed that your fork provides the AIC based on the Fisher’s C statistic and d-sep test (AIC=C+2K), rather than the default AIC based on the log-likelihood chi-squared (and I guess this relates to the LLchisq() issue above). For model comparison/selection, I am therefore manually calculating the default AIC.
Anyway, speak soon and thanks again!

**Comment by jslefche** [comment]
Appears to be fixed in 2.2.1

**Comment by Canderson156** [comment]
Seems like this is still an issue that hasn't been resolved? I tried the reprex at the beginning of the thread and it still doesn't work on piecewisesem 2.2.1. The fix here seems to be a forked version, which is from 10 months ago, so I'm hesitant to use this version since I'm assuming the main package has been updated since then. Could the updated class name be added to the main version of the package?
